ArrayIndexOutOfBoundsException in StageQueueImpl.findShortestQueueIndex

During a long-running test I encountered the following exception on the server (this was when using four clients on a single enterprise server):

com.tc.exception.TCRuntimeException: Uncaught exception in stage
        at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:211)
Caused by: java.lang.ArrayIndexOutOfBoundsException: -58
        at com.tc.async.impl.StageQueueImpl.findShortestQueueIndex(StageQueueImpl.java:197)
        at com.tc.async.impl.StageQueueImpl.getSourceQueueFor(StageQueueImpl.java:213)
        at com.tc.async.impl.StageQueueImpl.addMultiThreaded(StageQueueImpl.java:141)
        at com.tc.objectserver.entity.RequestProcessor.scheduleRequest(RequestProcessor.java:86)
        at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.start(ManagedEntityImpl.java:1019)
        at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.access$000(ManagedEntityImpl.java:976)
        at com.tc.objectserver.entity.ManagedEntityImpl.scheduleInOrder(ManagedEntityImpl.java:279)
        at com.tc.objectserver.entity.ManagedEntityImpl.processInvokeRequest(ManagedEntityImpl.java:308)
        at com.tc.objectserver.entity.ManagedEntityImpl.addRequestMessage(ManagedEntityImpl.java:192)
        at com.tc.objectserver.handler.ProcessTransactionHandler.addMessage(ProcessTransactionHandler.java:294)
        at com.tc.objectserver.handler.ProcessTransactionHandler.access$400(ProcessTransactionHandler.java:75)
        at com.tc.objectserver.handler.ProcessTransactionHandler$2.handleEvent(ProcessTransactionHandler.java:140)
        at com.tc.objectserver.handler.ProcessTransactionHandler$2.handleEvent(ProcessTransactionHandler.java:122)
        at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
        at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)

[terracotta-server.txt](https://github.com/Terracotta-OSS/terracotta-core/files/765312/terracotta-server.txt)

--- a/common/src/main/java/com/tc/async/impl/StageQueueImpl.java
+++ b/common/src/main/java/com/tc/async/impl/StageQueueImpl.java
@@ -69,7 +69,18 @@ public class StageQueueImpl<EC> implements Sink<EC> {
     this.sourceQueues = new SourceQueueImpl[queueCount];
     createWorkerQueues(queueCount, queueFactory, queueSize, stageName);
   }
-
+  
+  @SuppressWarnings("unchecked")  // for tests
+  StageQueueImpl(int queueCount, QueueFactory<ContextWrapper<EC>> queueFactory,
+                        TCLoggerProvider loggerProvider, String stageName, int queueSize, int fc) {
+    Assert.eval(queueCount > 0);
+    this.logger = loggerProvider.getLogger(Sink.class.getName() + ": " + stageName);
+    this.stageName = stageName;
+    this.sourceQueues = new SourceQueueImpl[queueCount];
+    this.fcheck = fc;
+    createWorkerQueues(queueCount, queueFactory, queueSize, stageName);
+  }
+  
   private void createWorkerQueues(int queueCount, QueueFactory<ContextWrapper<EC>> queueFactory, int queueSize, String stage) {
     StageQueueStatsCollector statsCollector = new NullStageQueueStatsCollector(stage);
     BlockingQueue<ContextWrapper<EC>> q = null;
@@ -185,7 +196,7 @@ public class StageQueueImpl<EC> implements Sink<EC> {
   private volatile int fcheck = 0;
 //  TODO:  Way too busy. need a better way
   private int findShortestQueueIndex() {
-      int stop = (fcheck++) % this.sourceQueues.length;
+      int stop = Math.abs(fcheck++) % this.sourceQueues.length;
       int pointer = stop+1;
       int min = Integer.MAX_VALUE;
       int can = -1;

--- a/common/src/test/java/com/tc/async/impl/StageQueueImplTest.java
+++ b/common/src/test/java/com/tc/async/impl/StageQueueImplTest.java
@@ -21,6 +21,7 @@ package com.tc.async.impl;
 import com.tc.async.api.MultiThreadedEventContext;
 import com.tc.logging.DefaultLoggerProvider;
 import com.tc.logging.TCLoggerProvider;
+import com.tc.util.Assert;
 import com.tc.util.concurrent.QueueFactory;
 import java.util.ArrayList;
 import java.util.List;
@@ -132,5 +133,31 @@ public class StageQueueImplTest {
     instance.addMultiThreaded(context1);
 //  tests specific implementation.  test expectation
     assertEquals(cxts.get(rand % cxts.size()).poll(), context1);
+  }
+  
+  @Test
+  public void testShortestRollover() {
+    TCLoggerProvider logger = new DefaultLoggerProvider();
+    final List<BlockingQueue<Object>> cxts = new ArrayList<BlockingQueue<Object>>();
+    
+    QueueFactory<ContextWrapper<Object>> context = mock(QueueFactory.class);
+    when(context.createInstance(Matchers.anyInt())).thenAnswer(new Answer<BlockingQueue<Object>>() {
+
+      @Override
+      public BlockingQueue<Object> answer(InvocationOnMock invocation) throws Throwable {
+        BlockingQueue<Object> queue = new ArrayBlockingQueue<Object>((Integer)invocation.getArguments()[0]);
+        cxts.add(queue);
+        return queue;
+      }
+    
+    });
+    StageQueueImpl impl = new StageQueueImpl(1, context, logger, "mock", 16, Integer.MAX_VALUE);
+    MultiThreadedEventContext cxt = mock(MultiThreadedEventContext.class);
+    when(cxt.getSchedulingKey()).thenReturn(null);
+ // hit int max
+    impl.addMultiThreaded(cxt);
+ // rollover
+    impl.addMultiThreaded(cxt);
+    Assert.assertEquals(cxts.get(0).size(), 2);
   }
 }

